scripts/http-jsonp-detection.nse

```diff
@@ -0,0 +1,174 @@
+local http = require "http"
+local shortport = require "shortport"
+local stdnse = require "stdnse"
+local string = require "string"
+local json = require "json"
+local url = require "url"
+local httpspider = require "httpspider"
+
+description = [[
+Attempts to discover JSONP endpoints in web servers. JSONP endpoints can be
+used to bypass Same-origin Policy restrictions in web browsers.
+
+The script searches for callback functions in the response to detect JSONP
+endpoints. It also tries to determine callback function through URL(callback 
+function may be fully or partially controllable from URL) and also tries to 
+bruteforce the most common callback variables through the URL.
+
+References : https://securitycafe.ro/2017/01/18/practical-jsonp-injection/
+
+]]
+
+---
+-- @usage
+-- nmap -p 80 --script http-jsonp-detection <target>
+--
+-- @output
+-- 80/tcp open  http    syn-ack
+-- | http-jsonp-detection: 
+-- |   JSONP endpoint found at /rest/contactsjp.php. Function name is callback
+-- |_  Callback function is completely controllable from the URL
+-- 
+--
+-- @xmloutput
+-- <table>
+-- <elem>JSONP endpoint found at /rest/contactsjp.php. Function name is callback</elem>
+-- <elem>Callback function is completely controllable from the URL</elem>
+-- </table>
+-- 
+-- @args http-jsonp-detection.path The URL path to request. The default path is "/".
+--- 
+
+author = {"Vinamra Bhatia"}
+license = "Same as Nmap--See https://nmap.org/book/man-legal.html"
+categories = {""} --to be figured out!
+
+portrule = shortport.port_or_service({80,443}, "http", "tcp")
+
+local callbacks = {"cb", "jsonp", "jsonpcallback", "jcb", "call"}
+
+--Checks the body and returns if valid json data is present in callback function
+local checkjson = function(body)
+   
+  local func, json_data
+  _, _, func, json_data = string.find(body, "(%S+)%((.*)%)")  --Check this!
+
+  --Check if the json_data is valid
+  --If valid, we have a JSONP endpoint with func as the function name
+
+  local status, json = json.parse(json_data)
+  return status, func
+end
+
+--Checks if the callback function is controllable from URL
+local callback_url = function(host, port, target)
+  local path, response, report
+  path = target .. "?callback=testing"
+  response = http.get(host, port, path)
+  if response and response.body and response.status==200 then
```

Comment:
response.status might be nil and an error will be thrown when you attempt to compare response.status == 200